### DISCUSSION TITLE
Registering a module with arch-specific packages

### DISCUSSION THREAD

**Discussion by phaedon** [bazelbuild/bazel-central-registry#792] [Q&A]
Hi, I'd like to add a new module for Embree (Intel's ray tracing engine) to the registry.

Best way to do this seems to be to use one of the pre-compiled shared libraries, available at
https://github.com/embree/embree/releases/tag/v4.1.0
(and not building it from source)

However, I am not sure how to set up the BCR config, given that each BCR module version subdir has exactly one source.json file with one url field. Ideally I'd want to be able to either:
 - set the version name to "4.1.0.x86_64.linux" but this is not permitted because of the regex in bcr_validation.py
 - define multiple URLs in embree/4.1.0/source.json -- one for linux, one for mac osx, and one for windows -- but I think this is not possible

As a workaround I could name the top-level module "embree-x86_64-linux" but this seems like it would quickly pollute the top-level namespace of the BCR and just a bad idea. It's almost better to name it "embree" but to only support linux for now.

What's the recommended solution to this?

**Comment by fmeum** [comment]
You could have a single `embree` module with an internal module extension that defines `http_file`s for each architecture and then contains a single `cc_import` that `select`s the prebuilt library from the correct repo depending on the target architecture.

This has the advantage that the architecture decision can be handled with regular Bazel tools (platforms) and doesn't play a rule during dependency resolution. Lazy repo fetching means that only the relevant arch will be downloaded.

**Reply by phaedon** [reply]
Thanks for the quick reply. My initial attempt doesn't work:
https://github.com/bazelbuild/bazel-central-registry/compare/main...phaedon:bazel-central-registry:embree

There are two issues that I'd love some help on:
1. I'm using http_archive, since these are tar/zip files, but apparently I can't use that outside of WORKSPACE. How do I use http_file and unzip the contents? I couldn't find a relevant example.
2. I have to specify some sort of url in source.json and that apparently is usually a tar file, so I left it pointing to the linux version. Is there a way to remove this? I think I need the file in order to provide the sha256 codes for the patch files, but I tried setting "url" to null and it complained.

**Reply by fmeum** [reply]
> 1. I'm using http_archive, since these are tar/zip files, but apparently I can't use that outside of WORKSPACE. How do I use http_file and unzip the contents? I couldn't find a relevant example.

You are right to use `http_archive` in this case. This is an example for how you use a module extension to load a few `http_archive`s: https://github.com/bazelbuild/rules_go/blob/98165a6230a0702aac310d6eb764477c00504e2b/MODULE.bazel#L13


> 2. I have to specify some sort of url in source.json and that apparently is usually a tar file, so I left it pointing to the linux version. Is there a way to remove this? I think I need the file in order to provide the sha256 codes for the patch files, but I tried setting "url" to null and it complained.

This is an interesting case as all the files of your module come from patches. You could reference an empty archive or add your files to a new git repository you marathon. The latter would allow you to get rid of the patches.
